hubblestack/extmods/fileserver/s3fs.py

```diff
@@ -0,0 +1,742 @@
+            if bucket_name not in found:
+                found[bucket_name] = True
+                ret.append({bucket_name: filepaths})
+            else:
+                for bucket in ret:
+                    if bucket_name in bucket:
+                        bucket[bucket_name] += filepaths
+                        break
+    return ret
+
+
+def _find_dirs(metadata):
+    '''
+    Looks for all the directories in the S3 bucket cache metadata.
+
+    Supports trailing '/' keys (as created by S3 console) as well as
+    directories discovered in the path of file keys.
+    '''
+
+    ret = []
+    found = {}
+
+    for bucket_dict in metadata:
+        for bucket_name, data in six.iteritems(bucket_dict):
+            dirpaths = set()
+            for path in [k['Key'] for k in data]:
+                prefix = ''
+                for part in path.split('/')[:-1]:
+                    directory = prefix + part + '/'
+                    dirpaths.add(directory)
+                    prefix = directory
+            if bucket_name not in found:
+                found[bucket_name] = True
+                ret.append({bucket_name: list(dirpaths)})
+            else:
+                for bucket in ret:
+                    if bucket_name in bucket:
+                        bucket[bucket_name] += list(dirpaths)
+                        bucket[bucket_name] = list(set(bucket[bucket_name]))
+                        break
+    return ret
+
+
+def _find_file_meta(metadata, bucket_name, saltenv, path):
+    '''
+    Looks for a file's metadata in the S3 bucket cache file
+    '''
+    env_meta = metadata[saltenv] if saltenv in metadata else {}
+    bucket_meta = {}
+    for bucket in env_meta:
+        if bucket_name in bucket:
+            bucket_meta = bucket[bucket_name]
+    files_meta = list(list(filter((lambda k: 'Key' in k), bucket_meta)))
+
+    for item_meta in files_meta:
+        if 'Key' in item_meta and item_meta['Key'] == path:
+            try:
+                # Get rid of quotes surrounding md5
+                item_meta['ETag'] = item_meta['ETag'].strip('"')
+            except KeyError:
+                pass
+            return item_meta
+
+
+def _get_buckets():
+    '''
+    Return the configuration buckets
+    '''
+
+    return __opts__['s3.buckets'] if 's3.buckets' in __opts__ else {}
+
+
+def _get_file_from_s3(metadata, saltenv, bucket_name, path, cached_file_path):
+    '''
+    Checks the local cache for the file, if it's old or missing go grab the
+    file from S3 and update the cache
+    '''
+    key, keyid, service_url, verify_ssl, kms_keyid, location, path_style, https_enable = _get_s3_key()
+
+    # check the local cache...
+    if os.path.isfile(cached_file_path):
+        file_meta = _find_file_meta(metadata, bucket_name, saltenv, path)
+        if file_meta:
+            file_etag = file_meta['ETag']
+
+            if file_etag.find('-') == -1:
+                file_md5 = file_etag
```

Comment:
Wouldn't this fail if the Etag is not md5?
It may not be as described here: https://docs.aws.amazon.com/AmazonS3/latest/API/RESTCommonResponseHeaders.html and it is determined by how the file has been uploaded. 
The largest file we have is already almost 5MB so it's potentially getting close. Worse, if this fails every time, there would never be a cache hit, and we'd be downloading the largest file _every_ time.

Reply:
Let me look. This isn't code that I touched, you can see my actual diff here: https://github.com/saltstack/salt/pull/47140

Even if we downloaded it every time, "every time" is only hourly, so not a huge deal, but I'll still see if I can fix it.

Reply:
Depends on how many clients per hour and the bandwidth. Adds up.

Reply:
I added a documentation note describing how to avoid multipart uploads.